consensus/dolev_strong.go

```diff
@@ -0,0 +1,310 @@
+	"github.com/spacemeshos/go-spacemesh/consensus/pb"
+	"github.com/spacemeshos/go-spacemesh/log"
+	//"reflect"
+	"bytes"
+	dsCfg "github.com/spacemeshos/go-spacemesh/consensus/config"
+	"github.com/spacemeshos/go-spacemesh/crypto"
+	"time"
+	"errors"
+	"fmt"
+)
+
+// DolevStrongConsensus is an implementation of a byzanteen agreement protocol
+type DolevStrongConsensus struct {
+	activeInstances map[string]CAInstance
+	ingressChannel  chan OpaqueMessage
+	net             NetworkConnection
+	dsConfig        dsCfg.Config
+	privateKey      crypto.PrivateKey
+	publicKey       crypto.PublicKey
+	timer           Timer
+	startTime       time.Time
+	abortInstance   chan struct{}
+	abortListening  chan struct{}
+}
+
+// DolevStrongInstance is a struct holding state for a single instance of a dolev strong agreement protocol as a receiver
+type DolevStrongInstance struct {
+	value                   []byte
+	ds                      *DolevStrongConsensus
+	abortProcessingInstance bool
+}
+
+// StartInstance starts an agreement protocol instance
+func (impl *DolevStrongConsensus) StartInstance(msg OpaqueMessage) []byte {
+	impl.startTime = time.Now()
+	go impl.SendMessage(msg)
+	go impl.StartListening()
+	impl.waitForConsensus()
+	return msg.Data()
+}
+
+// NewDSC creates a new instance of dolev strong agreement protocol
+func NewDSC(conf dsCfg.Config,
+	timer Timer,
+	nodes []string,
+	network NetworkConnection,
+	pubKey crypto.PublicKey,
+	privKey crypto.PrivateKey) (Algorithm, error) {
+
+	dsc := &DolevStrongConsensus{
+		activeInstances: make(map[string]CAInstance),
+		ingressChannel:  make(chan OpaqueMessage),
+		publicKey:       pubKey,
+		privateKey:      privKey,
+		net:             network,
+		dsConfig:        conf,
+		timer:           timer,
+		startTime:       time.Now(),
+		abortInstance:   make(chan struct{}, 1),
+		abortListening:  make(chan struct{}, 1),
+	}
+	for _, node := range nodes {
+		dsc.activeInstances[node] = NewCAInstance(dsc)
+	}
+
+	return dsc, nil
+}
+
+func (impl *DolevStrongConsensus) handleMessage(message OpaqueMessage) error{
+	msg := &pb.ConsensusMessage{}
+	proto.Unmarshal(message.Data(), msg)
+	pk, err := crypto.NewPublicKey(msg.Msg.AuthPubKey)
+	if err != nil {
+		log.Error("cannot parse public key from message: ", msg.Msg.AuthPubKey)
+		return err
+	}
+	nodeID := pk.String()
+	var ci CAInstance
+	if val, ok := impl.activeInstances[nodeID]; !ok {
+		ci = NewCAInstance(impl)
+		impl.activeInstances[nodeID] = ci
+	} else {
+		ci = val
+	}
+	return ci.ReceiveMessage(msg)
+}
+
+// StartListening starts listening for agreement protocol messages from other instances running in the network
+func (impl *DolevStrongConsensus) StartListening() error {
+	impl.ingressChannel = impl.net.RegisterProtocol("DolevStrong")
+	log.Info("started listening")
+	for {
+		select {
+		case message := <-impl.ingressChannel:
+			err := impl.handleMessage(message)
+			if err != nil {
+				log.Error("Protocol encountered error: %v", err)
```

Comment:
shouldn't it return with the error?

Reply:
If this method should ignore message errors (make sense to me) then it shouldn't return error at all, note that currently it will always return nil